crates/sui/src/upgrade_compatibility.rs

```diff
@@ -353,6 +359,56 @@ impl CompatibilityMode for CliCompatibilityMode {
     }
 }
 
+/// Generates an enum Category along with individual enum for each individual category
+/// and impls into diagnostic info for each category.
+macro_rules! upgrade_codes {
+    ($($cat:ident: [
+        $($code:ident: { msg: $code_msg:literal }),* $(,)?
+    ]),* $(,)?) => {
+        #[derive(PartialEq, Eq, Clone, Copy, Debug, Hash, PartialOrd, Ord)]
+        #[repr(u8)]
+        pub enum Category {
+            #[allow(dead_code)]
+            ZeroPlaceholder,
+            $($cat,)*
+        }
+
+        $(
+            #[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
+            #[repr(u8)]
+            pub enum $cat {
+                #[allow(dead_code)]
+                ZeroPlaceholder,
+                $($code,)*
+            }
+
+            // impl into diagnostic info
+            impl Into<DiagnosticInfo> for $cat {
+                fn into(self) -> DiagnosticInfo {
+                    match self {
+                        Self::ZeroPlaceholder =>
+                            panic!("do not use placeholder error code"),
+                        $(Self::$code => custom(
+                            "COMPATIBILITY",
+                            Severity::NonblockingError,
+                            Category::$cat as u8,
+                            self as u8,
+                            $code_msg,
+                        ),)*
+                    }
+                }
+            }
+        )*
+    };
+}
+
+// Used to generate diagnostics primary labels for upgrade compatibility errors.
+upgrade_codes!(
+    Declarations: [
+        PublicMissing: { msg: "missing public declaration" },
+    ],
+);
```

Comment:
I think in general this is fine, but the syntatic structure here was designed for a large amount of categories+codes with a lot of per-code tweaking, which isn't something we are doing here. 

We might want something a tad simpler like 
external-crates/move/crates/move-compiler/src/linters/mod.rs

But at the end of the day its up to you :)

Reply:
It's probably overkill but I find it to be a relatively clean way to do it. Each category has its own set of codes which made the most sense to me.